admin/gv_mail.php

```diff
@@ -4,421 +4,382 @@
-    $audience_select = get_audience_sql_query($_POST['customers_email_address'], 'email');
-    $mail = $db->Execute($audience_select['query_string']);
-    $mail_sent_to = $audience_select['query_name'];
-    if ($_POST['email_to']) {
-      $mail_sent_to = $_POST['email_to'];
-    }
-
-    $from = zen_db_prepare_input($_POST['from']);
-    $subject = zen_db_prepare_input($_POST['subject']);
-    $recip_count=0;
-
-    // set time-limit for processing to 5 minutes... if allowed by PHP configuration
-    zen_set_time_limit(600);
-
-    while (!$mail->EOF) {
+}
 
-      $id1 = zen_create_coupon_code($mail->fields['customers_email_address']);
-      $insert_query = $db->Execute("insert into " . TABLE_COUPONS . "
+$_POST['amount'] = !empty($_POST['amount']) ? preg_replace('/[^0-9.%]/', '', $_POST['amount']) : 0;
+$_POST['amount'] = abs($_POST['amount']);
+$action = isset($_GET['action']) ? zen_db_prepare_input($_GET['action']) : '';
+if ($action != '') {
+    switch ($action) {
+        case ('preview'):
+            $error = 0;
+            if (empty($_POST['customers_email_address']) && empty($_POST['email_to'])) {
+                $messageStack->add(ERROR_NO_CUSTOMER_SELECTED, 'error');
+                $error++;
+            }
+            if (empty($_POST['subject'])) {
+                $messageStack->add(ERROR_NO_SUBJECT, 'error');
+                $error++;
+            }
+            if (empty($_POST['amount']) || $_POST['amount'] <= 0) {
+                $messageStack->add(ERROR_NO_AMOUNT_ENTERED, 'error');
+                $error++;
+            }
+            if ($error > 0) {
+                $action = $_GET['action'] = '';
+            }
+            break;
+
+        case ('send_email_to_user'):
+            if (isset($_POST['back'])) {
+                break;
+            }
+            if (isset($_POST['cancel'])) {
+                unset($_POST);
+                break;
+            }
+
+            $from_name = $from_email_address = zen_db_prepare_input($_POST['from']);
+            $subject = zen_db_prepare_input($_POST['subject']);
+            $recip_count = 0;
+
+            if ($_POST['email_to']) {
+                $mail_sent_to = zen_db_prepare_input($_POST['email_to']);
+                if (!empty($_POST['email_to_name'])) {
+                    $mail_sent_to_names = explode(' ', zen_db_prepare_input($_POST['email_to_name']), 2);
+                    $customers_firstname = $mail_sent_to_names[0];
+                    $customers_lastname = $mail_sent_to_names[1];
+                } else {
+                    $customers_firstname = '';
+                    $customers_lastname = TEXT_CUSTOMER;
+                }
+                $mail = (object)array('fields' => array('customers_firstname' => $customers_firstname, 'customers_lastname' => $customers_lastname, 'customers_email_address' => $mail_sent_to));

```

Comment:
Does this `(object)` typecasting still need to be here? I don't see where `$mail` is used as an object later, since it's being used in `foreach` instead of the older `while...MoveNext ... ->fields` syntax.

Reply:
agreed (now I've just looked it up..)